scio-core/src/main/scala/com/spotify/scio/io/BinaryIO.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2019 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.io
+
+import java.io.OutputStream
+import java.nio.channels.{Channels, WritableByteChannel}
+
+import com.spotify.scio.ScioContext
+import com.spotify.scio.values.SCollection
+import org.apache.beam.sdk.io._
+import org.apache.beam.sdk.transforms.PTransform
+import org.apache.beam.sdk.values.PCollection
+
+import scala.concurrent.Future
+
+/**
+ * A ScioIO class for writing raw bytes to files.
+ * Like TextIO, but without newline delimiters and operating over Array[Byte] instead of String.
+ * @param path a path to write to.
+ */
+final case class BinaryIO(path: String) extends ScioIO[Array[Byte]] {
+
+  override type ReadP = Nothing
+  override type WriteP = BinaryIO.WriteParam
+  override final val tapT = EmptyTapOf[Array[Byte]]
+
+  override def testId: String = s"BinaryIO($path)"
+
+  override def read(sc: ScioContext, params: ReadP): SCollection[Array[Byte]] =
+    throw new IllegalStateException("BinaryIO is read-only")
+
+  override def write(data: SCollection[Array[Byte]], params: WriteP): Future[Tap[Nothing]] = {
+    data.applyInternal(BinaryIO.writeBinary(path, params))
+    Future.successful(EmptyTap)
+  }
+
+  override def tap(params: Nothing): Tap[Nothing] = EmptyTap
+}
+
+object BinaryIO {
+  final case class WriteParam(suffix: String = ".bin",
+                              numShards: Int = 0,
+                              compression: Compression = Compression.UNCOMPRESSED)
+
+  private final class BytesSink extends FileIO.Sink[Array[Byte]] {
+    @SuppressWarnings(Array("org.wartremover.warts.Var"))
+    @transient private var channel: OutputStream = _
+
+    def open(channel: WritableByteChannel): Unit =
+      this.channel = Channels.newOutputStream(channel)
+
+    def flush(): Unit = {
+      if (this.channel == null) {
+        throw new IllegalStateException("Trying to flush a BytesSink that has not been opened")
+      }
+
+      this.channel.flush()
+    }
+
+    def write(datum: Array[Byte]): Unit = {
+      if (this.channel == null) {
+        throw new IllegalStateException("Trying to write to a BytesSink that has not been opened")
+      }
+
+      this.channel.write(datum)
+    }
+  }
+
+  def writeBinary(
```

Comment:
@psobot I think I would move this logic to the `BinaryIO.write` it would be consistent with all IO's which normally will expose read/write